src/main/java/net/dv8tion/jda/bot/entities/impl/DefaultShardManagerImpl.java

```diff
@@ -0,0 +1,336 @@
+                LOG.warn("Hit the login ratelimit while creating new JDA instances, maybe try raising the login backoff");
+            }
+            catch (LoginException e)
+            {
+                // this can only happen if the token has been changed
+                // in this case the ShardManager will just shutdown itself as there currently is no way of hot-swapping the token on a running JDA instance.
+                LOG.warn("The token has been invalidated and the ShardManager will shutdown!");
+                this.shutdown();
+            }
+            catch (final Exception e)
+            {
+                if (api != null)
+                    api.shutdown();
+            }
+        }, 5000 + this.backoff, 5000 + this.backoff, TimeUnit.MILLISECONDS); // 5s for ratelimit + backoff for safety
+
+        if (this.shutdownHook != null)
+            Runtime.getRuntime().addShutdownHook(this.shutdownHook);
+    }
+
+    @Override
+    public void removeEventListener(final Object... listeners)
+    {
+        this.listeners.removeAll(Arrays.asList(listeners));
+        ShardManager.super.removeEventListener(listeners);
+    }
+
+    @Override
+    public void restart()
+    {
+        for (final int shardId : this.shards.getMap().keys())
+        {
+            final JDA jda = this.shards.getMap().remove(shardId);
+            if (jda != null)
+            {
+                jda.shutdown();
+                this.queue.offer(shardId);
+            }
+        }
+    }
+
+    @Override
+    public void restart(final int shardId)
+    {
+        Checks.notNegative(shardId, "shardId");
+        Checks.check(shardId < this.shardsTotal, "shardId must be lower than shardsTotal");
+
+        final JDA jda = this.shards.getMap().remove(shardId);
+        if (jda != null)
+            jda.shutdown();
+
+        this.queue.add(shardId);
+    }
+
+    @Override
+    public void shutdown()
+    {
+        if (this.shutdown.getAndSet(true))
+            return; // shutdown has already been requested
+
+        if (this.worker != null && !this.worker.isDone())
+            this.worker.cancel(true);
+
+        if (this.shutdownHook != null)
+        {
+            try
+            {
+                Runtime.getRuntime().removeShutdownHook(this.shutdownHook);
+            }
+            catch (final Exception ignored) {}
+        }
+
+        this.executor.shutdown();
+
+        if (this.shards != null)
+        {
+            for (final JDA jda : this.shards)
+                jda.shutdown();
+        }
+    }
+
+    @Override
+    public void shutdown(final int shardId)
+    {
+        final JDA jda = this.shards.getMap().remove(shardId);
+        if (jda != null)
+            jda.shutdown();
+    }
+
+    @Override
+    public void start(final int shardId)
+    {
+        Checks.notNegative(shardId, "shardId");
+        Checks.check(shardId < this.shardsTotal, "shardId must be lower than shardsTotal");
+        this.queue.add(shardId);
+    }
```

Comment:
Does this allow starting a shard multiple times?

Reply:
It does allow adding an Id multiple times to the queue, but if the building thread polls from the queue it checks if the shard already exists and just moves on to the next queue entry if so.

Reply:
Can you document that case if it isn't already documented

Reply:
Will do that